tree/treeplayer/src/TSimpleAnalysis.cxx

```diff
@@ -219,41 +239,125 @@ bool TSimpleAnalysis::Run()
       fTreeName = ExtractTreeName(fInputFiles[0]);
    if (fTreeName.empty())  // No tree name found
       return false;
+   return true;
+}
 
-   // Do the chain of the fInputFiles
-   TChain chain(fTreeName.c_str());
-   for (const std::string& inputfile: fInputFiles)
-      chain.Add(inputfile.c_str());
-
-   // Sanity check that we can open the first file
-   int errValue = chain.LoadTree(0);
-   if (errValue < 0) {
-      ::Error("TSimpleAnalysis::Analyze",
-              "The chain is not correctly set up, chain.LoadTree(0) returns %d", errValue);
+////////////////////////////////////////////////////////////////////////////////
+/// Execute all the TChain::Draw() as configured and stores the output histograms.
+/// Returns true if the analysis succeeds.
+
+bool TSimpleAnalysis::Run()
+{
+   if (!SetTreeName())
       return false;
-   }
 
    TFile ofile(fOutputFile.c_str(), "RECREATE");
    if (ofile.IsZombie()) {
-      ::Error("TSimpleAnalysis::Analyze", "Impossible to create %s", fOutputFile.c_str());
+      ::Error("TSimpleAnalysis::Run", "Impossible to create %s", fOutputFile.c_str());
       return false;
    }
 
-   // Save the histograms into the output file
-   for (const auto &histo : fHists) {
-      const std::string& expr = histo.second.first;
-      const std::string& histoName = histo.first;
-      const std::string& cut = histo.second.second;
+   // Possible return values of TChain::LoadTree()
+   static const char* errors[] {
+      "all good", // 0
+         "empty chain", // -1
+         "invalid entry number", // -2
+         "cannot open the file", // -3
+         "missing tree", // -4
+         "internal error" // -5
+         };
+
+   auto generateHisto = [&](const std::pair<TChain*, TDirectory*>& job){
+      TChain* chain = job.first;
+      TDirectory* taskDir = job.second;
+      taskDir->cd();
+      std::vector<TH1F *> vPtrHisto;
+      for (const auto &histo : fHists) {
+         const std::string& expr = histo.second.first;
+         const std::string& histoName = histo.first;
+         const std::string& cut = histo.second.second;
+
+         chain->Draw((expr + ">>" + histoName).c_str(), cut.c_str(), "goff");
+         TH1F *ptrHisto = (TH1F*)taskDir->Get(histoName.c_str());
+         int errValue;
+         std::string errFile;
+         if (!checkChainLoadResult(chain, errValue, errFile)) {
+            ::Error("TSimpleAnalysis::Run",
+                    "Load failure in file %s: %s", errFile.c_str(), errors[-errValue]);
+            return std::vector<TH1F *>();
+         }
+         vPtrHisto.emplace_back(ptrHisto);
+      }
+      return vPtrHisto;
+   };
+
+#if 1
+
```

Comment:
Please add this comment:
```
The MT version is currently disabled because reading emulated objects triggers a lock for every object read. This in turn increases the run time way beyond the serial case.
```